README.md
archlinux_macbookretina puppet module
=====================================

[![Project Status: Active - The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/0.1.0/active.svg)](http://www.repostatus.org/#active)

Puppet module and accompanying documentation to install/setup Arch linux on a MacBook Pro Retina using puppet4.

This is the puppet module I use to manage my shiny new MacBook Pro Retina (mine is a ``MacBookPro11,4``).

__Important Notice:__ It is _highly_ recommended that you only run this module on a brand new system;
it makes some opinionated choices that may cause problems on existing systems.

Requirements
------------

* A MacBook Pro Retina (currently tested with Mid-2015 "11,4" model) with a __brand new install__ of Arch Linux
* Puppet4
* A minimal install, ideally following the instructions in my [workstation-bootstrap](https://github.com/jantman/workstation-bootstrap#arch-linux) repository.
* This module is only tested alongside my [archlinux_workstation](https://github.com/jantman/puppet-archlinux-workstation) module.
* A pacman repo providing AUR packages: [macfanctld](https://aur.archlinux.org/packages/macfanctld), [bcwc-pcie-dkms](https://aur.archlinux.org/packages/bcwc-pcie-dkms/) and [bcwc-pcie-firmware](https://aur.archlinux.org/packages/bcwc-pcie-firmware/); you can use [jantman/archlinux_workstation archlinux_workstation::repos::jantman](https://github.com/jantman/puppet-archlinux-workstation/blob/master/manifests/repos/jantman.pp) for that.

Hardware Support Status
------------------------

### Mid-2015 MacBookPro11,4 as of 2016-10-25 (4.8.4-1 kernel)

#### Working

Or mostly-working:

* __Touchpad__ using Kernel 4.2+ and [xf86-input-synaptics](https://www.archlinux.org/packages/?name=xf86-input-synaptics); works for tap-to-click, drag, two-finger scroll, and 2- or 3-finger taps (sometimes) for different mouse buttons.
* __Networking__ on MacBookPro 11,4
  * USB ethernet adapter A1277 works out-of-the-box
  * BCM43602 AirPort Extreme (14e4:43ba) works with kernel built-in [brcmfmac](https://wireless.wiki.kernel.org/en/users/drivers/brcm80211) driver, autodetected, both 2.4GHz and 5GHz.
* __Sound__ - Works. Under KDE/Phonon, needed to unmute/enable the "Built-in Audio Analog Stereo" (detected "Built-in Audio Digital Stereo (HDMI)" as default).
* __Video__ - video works with the proprietary nvidia driver, the default in this module.
  * __External Displays__ - Tested OK using both direct HDMI and Thunderbolt to HDMI (1 or 2 external monitors); works seamlessly.
  * __Display/Desktop Scaling__ - This can be fixed within KDE:
    * System Settings -> Fonts: check off "Force fonts DIP" and set to 144
    * System Settings -> Icons -> "Advanced" tab: set them all to 48
    * Click the menu button on the far right edge of the Panel, then drag the "Height" box up until the scale/size looks good
  * __Screen Backlight Adjustment__ - works using ``/sys/class/backlight/acpi_video0/brightness``, the sliders on 'KDE5 System Settings -> Energy Saving' or the keyboard function keys.
* __Hibernate / Suspend to Disk__ - doesn't wake up without long hold of power button & then turn back on. Session resumes once that's done.
* __SD Card Reader__ - Working out of the box.
* __Fans__ - [macfanctld (AUR)](https://aur.archlinux.org/packages/macfanctld)
* __Webcam__ - Per [Arch Wiki](https://wiki.archlinux.org/index.php/MacBookPro11,x#Web_cam), there's a [project on GitHub](https://github.com/patjak/bcwc_pcie/) for a reverse-engineered driver. This works properly in Linux using the [bcwc-pcie-dkms](https://aur.archlinux.org/packages/bcwc-pcie-dkms/) and [bcwc-pcie-firmware](https://aur.archlinux.org/packages/bcwc-pcie-firmware/) AUR packages, though the color balance seems a bit off.

#### Broken

* __Suspend to RAM__ - doesn't wake up; long hold of power button & then turn back on gives a fresh boot.
* __Lid Close__ - suspends to ram and doesn't wake up
* __Bluetooth__ - Per [wiki](https://wiki.archlinux.org/index.php/MacBook#Bluetooth_2), fully supported as of kernel 4.4.0. Works with my bluetooth mouse.

#### Untested / To Do

* __Partially-complete__ - __SSD optimizations__ via sysctl settings, mount /dev/sda* noatime and discard (TRIM), use deadline scheduler on non-rotational disks
* __Power Saving__ - https://wiki.archlinux.org/index.php/MacBookPro11,x#Powersave and https://wiki.archlinux.org/index.php/Laptop_Mode_Tools
* [Laptop Mode Tools - ArchWiki](https://wiki.archlinux.org/index.php/Laptop_Mode_Tools) and/or  [TLP - ArchWiki](https://wiki.archlinux.org/index.php/TLP) *(started looking into this, very involved configuration and I don't really need it right now)*
* I'm left handed. Use udev/xorg to reverse buttons on USB mice but keep trackpad the same. See https://wiki.archlinux.org/index.php/All_Mouse_Buttons_Working http://www.smop.co.uk/blog/index.php/2010/02/15/udev-rules-for-logitech-g7-mouse/ or might be able to do this with udev triggering "xinput set-button-map"
* the stuff in [Maximizing Performance - ArchWiki](https://wiki.archlinux.org/index.php/Maximizing_Performance)
* implement the stuff in [Enhancing Arch Linux Stability - ArchWiki](https://wiki.archlinux.org/index.php/Enhancing_Arch_Linux_Stability)
* SMART/other SSD health check, with warnings if things go south
* look into replacing macfanctld with mbpfan / fan-control-daemon are two options that are less abrupt

Testing
-------

### Unit Tests

Unit tests should be run automatically by TravisCI. To run them locally, use ``bundle exec rake spec``. To run the full commit tests (validation, syntax, lint, rspec, etc.) use ``bundle exec rake test``.

1. Use whatever ruby version is specified in ``.ruby-version``
2. ``bundle install --path vendor``
3. ``bundle exec rake test``

### Acceptance Tests

Acceptance tests are done with [beaker](https://github.com/puppetlabs/beaker) and [beaker-rspec](https://github.com/puppetlabs/beaker-rspec). They currently use the "2017.07.01" version of the [ogarcia/archlinux-x64 Vagrant box](https://atlas.hashicorp.com/ogarcia/boxes/archlinux-x64/) for VirtualBox.

1. Use whatever ruby version is specified in ``.ruby-version``
2. ``bundle install --path vendor``
3. ``bundle exec rake spec_prep`` to setup the fixture modules that the specs will need
4. ``bundle exec rake beaker`` to run the tests.

See [beaker-rspec](https://github.com/puppetlabs/beaker-rspec) for environment variables to control the run. The important settings are:

* ``BEAKER_destroy=no`` - keep the VMs after the test run
* ``BEAKER_destroy=onpass`` - to keep the VMs around only after a test failure

Other References
----------------

* First, many thanks to [eli meister](https://twitter.com/elitmeister) for being the office linux-on-MBP retina guinea pig
* [MacBookPro Retina - ArchWiki](https://wiki.archlinux.org/index.php/MacBookPro_Retina)
* [MacBook - ArchWiki](https://wiki.archlinux.org/index.php/MacBook)
* [Installation Guide - ArchWiki](https://wiki.archlinux.org/index.php/Installation_Guide)
* [Beginners' Guide - ArchWiki](https://wiki.archlinux.org/index.php/Beginners%27_Guide)
* [Solid State Drives - ArchWiki](https://wiki.archlinux.org/index.php/Solid_State_Drives)
* [Laptop - ArchWiki](https://wiki.archlinux.org/index.php/Laptop)
* [Enhancing Arch Linux Stability - ArchWiki](https://wiki.archlinux.org/index.php/Enhancing_Arch_Linux_Stability)
* [Arch Linux System Maintenance - ArchWiki](https://wiki.archlinux.org/index.php/Arch_Linux_System_Maintenance)


CHANGELOG
2017-07-09 release 0.2.2
- Update for AUR package changes: bcwc-pcie-dkms -> bcwc-pcie-git and bcwc-firmware -> facetimehd-firmware

2017-01-08 release 0.2.1
- updated README with note that bluetooth now works
- many changes to testing (TravisCI, spec, beaker acceptance) and use new versions of dependencies (thanks to petems)
- fix ordering bug in archlinux_macbookretina::touchpad (thanks to petems)

2016-10-25 release 0.2.0
- ensure xf86-input-mtrack-git is absent, in favor of xf86-input-synaptics
- update README.md with updated statuses as of 2016-10-25 and 4.8.4-1 kernel
  - touchpad works, but some 3-finger-tap issues
  - built-in ``brcmfmac`` WiFi driver works for both 2.4GHz and 5GHz
  - backlight adjustment and other hotkeys work with keyboard function keys, using kernel built-in drivers
  - webcam works using AUR reverse-engineered driver
- add ``archlinux_macbookretina::webcam`` for reverse-engineered driver

2015-09-16 Release 0.1.0
- complete rewrite to be more reusable, standalone module, and targeted at puppet4

2014-03-15 Release 0.0.1
- initial module creation


spec/classes/ssd_spec.rb
require 'spec_helper'

describe 'archlinux_macbookretina::ssd' do
  let(:facts) {{
    :osfamily        => 'Archlinux',
    :operatingsystem => 'Archlinux',
    :productname     => 'MacBookPro11,4',
    # structured facts
    :os              => { 'family' => 'Archlinux' },
    :dmi             => { 'product' => { 'name' => 'MacBookPro11,4' } },
    :disks           => {
      'sda' => {
        'model' => "APPLE SSD SM0256",
        'size' => "233.76 GiB",
        'size_bytes' => 251000193024,
        'vendor' => "ATA"
      },
      'sdb' => {
        'model' => "SD Card Reader",
        'size' => "0 bytes",
        'size_bytes' => 0,
        'vendor' => "APPLE"
      }
    },
  }}

  describe "installs packages" do
    it { should compile.with_all_deps }

    it { should contain_class('archlinux_macbookretina::ssd') }
    it { should contain_file('/etc/udev/rules.d/60-schedulers.rules').with({
                                                                           :ensure => 'present',
                                                                           :owner  => 'root',
                                                                           :group  => 'root',
                                                                           :mode   => '0644',
                                                                           :source => 'puppet:///modules/archlinux_macbookretina/60-schedulers.rules',
                                                                           })
    }

    it { should_not contain_sysctl__value('vm.dirty_writeback_centisecs') }
    it { should_not contain_sysctl__value('vm.laptop_mode') }
    it { should_not contain_sysctl__value('vm.swappiness') }
    it { should_not contain_sysctl__value('vm.vfs_cache_pressure') }
  end
end


spec/classes/init_spec.rb
require 'spec_helper'

describe 'archlinux_macbookretina' do
  let(:facts) {{
    :osfamily        => 'Archlinux',
    :operatingsystem => 'Archlinux',
    :productname     => 'MacBookPro11,4',
    # structured facts
    :os              => { 'family' => 'Archlinux' },
    :dmi             => { 'product' => { 'name' => 'MacBookPro11,4' } },
  }}

  context 'supported operating systems' do
    describe "archlinux_macbookretina class with username parameter on Archlinux" do
      it { should compile.with_all_deps }
      it { should contain_class('archlinux_macbookretina') }
    end
  end

  context 'unsupported operating system' do
    describe 'archlinux_macbookretina class without any parameters on Solaris/Nexenta' do
      let(:facts) {{
                     :osfamily        => 'Solaris',
                     :operatingsystem => 'Nexenta',
                     :productname     => 'MacBookPro11,4',
                     # structured facts
                     :os              => { 'family' => 'Solaris', 'name' => 'Nexenta', },
                     :dmi             => { 'product' => { 'name' => 'MacBookPro11,4' } },
      }}

      it { expect { should contain_class('archlinux_macbookretina') }.to raise_error(Puppet::Error, /Nexenta not supported/) }
    end
    describe 'archlinux_macbookretina class without any parameters on CentOS' do
      let(:facts) {{
                     :osfamily        => 'RedHat',
                     :operatingsystem => 'CentOS',
                     :productname     => 'MacBookPro11,4',
                     # structured facts
                     :os              => { 'family' => 'RedHat', 'name' => 'CentOS', },
                     :dmi             => { 'product' => { 'name' => 'MacBookPro11,4' } },
      }}

      it { expect { should contain_class('archlinux_macbookretina') }.to raise_error(Puppet::Error, /CentOS not supported/) }
    end
    describe 'archlinux_macbookretina class without any parameters on Debian' do
      let(:facts) {{
                     :osfamily        => 'Debian',
                     :operatingsystem => 'debian',
                     :productname     => 'MacBookPro11,4',
                     # structured facts
                     :os              => { 'family' => 'Debian', 'name' => 'debian', },
                     :dmi             => { 'product' => { 'name' => 'MacBookPro11,4' } },
      }}

      it { expect { should contain_class('archlinux_macbookretina') }.to raise_error(Puppet::Error, /debian not supported/) }
    end
  end

  context 'supported hardware' do
    describe 'MacBookPro10,1' do
      let(:facts) {{
                     :osfamily        => 'Archlinux',
                     :operatingsystem => 'Archlinux',
                     :productname     => 'MacBookPro10,1',
                     # structured facts
                     :os              => { 'family' => 'Archlinux' },
                     :dmi             => { 'product' => { 'name' => 'MacBookPro10,1' } },
      }}
      it { should compile.with_all_deps }
      it { should contain_class('archlinux_macbookretina') }
    end
    describe 'MacBookPro11,4' do
      let(:facts) {{
                     :osfamily        => 'Archlinux',
                     :operatingsystem => 'Archlinux',
                     :productname     => 'MacBookPro11,4',
                     # structured facts
                     :os              => { 'family' => 'Archlinux' },
                     :dmi             => { 'product' => { 'name' => 'MacBookPro11,4' } },
      }}
      it { should compile.with_all_deps }
      it { should contain_class('archlinux_macbookretina') }
    end
  end

  context 'unsupported hardware' do
    describe 'To be filled by O.E.M.' do
      let(:facts) {{
                     :osfamily        => 'Archlinux',
                     :operatingsystem => 'Archlinux',
                     :productname     => 'To be filled by O.E.M.',
                     # structured facts
                     :os              => { 'family' => 'Archlinux' },
                     :dmi             => { 'product' => { 'name' => 'To be filled by O.E.M.' } },
      }}
      it { expect { should contain_class('archlinux_macbookretina') }.to raise_error(Puppet::Error, /To be filled by O\.E\.M\. is not a supported hardware productname/) }
    end
  end
  context 'parameters' do
    describe 'default' do
      let(:facts) {{
                     :osfamily        => 'Archlinux',
                     :operatingsystem => 'Archlinux',
                     :productname     => 'MacBookPro11,4',
                     # structured facts
                     :os              => { 'family' => 'Archlinux' },
                     :dmi             => { 'product' => { 'name' => 'MacBookPro11,4' } },
      }}
      it { should compile.with_all_deps }
      it { should contain_class('archlinux_macbookretina') }
      it { should contain_class('archlinux_macbookretina::wireless') }
      it { should contain_class('archlinux_macbookretina::videodriver') }
      it { should contain_class('archlinux_macbookretina::mac_keyboard') }
      it { should contain_class('archlinux_macbookretina::ssd') }
      it { should contain_class('archlinux_macbookretina::macfanctld') }
      it { should contain_class('archlinux_macbookretina::touchpad') }
    end
  end
end


manifests/makeinitpp.sh
#!/bin/bash
#
# script to re-make init.pp to include all other pp files in this directory
#

# space-separated list of classes we don't want to automatically include-
# either partially implemented classes, or highly specific ones that most
# people won't want.
BLACKLIST_CLASSES="pdnsd laptop_mode_tools firewall_pre firewall_post nouveau"

pushd `dirname $0` > /dev/null

(
cat <<EOF
# Class: archlinux_macbookretina
#
# Manage includes of other archlinux_macbookretina classes
# top-level class for archlinux_macbookretina ecosystem
#
# WARNING - WARNING - this file is generated by makeinitpp.sh. 
#   any local changes will be overwritten.
#
# Parameters:
#
# Actions:
#   - Include other classes
#
# Requires:
#
# Sample Usage:
#
class archlinux_macbookretina {
EOF
) > init.pp

for i in `ls -1 *.pp | grep -v "init.pp" | awk -F \. '{print $1}'`
do
	# skip anything that isnt a class (defines, etc.)
	if ! grep "^class archlinux_macbookretina::$i" $i.pp > /dev/null
	then
		continue
	fi

	summary=`awk '{if (NR==3) print $0}' $i.pp`

	# skip blacklisted classes
	if [[ "$BLACKLIST_CLASSES" == *"$i"* ]]
	then
	    echo "# BLACKLISTED in makeinitpp.sh: $i" >> init.pp
	    echo "#  $summary" >> init.pp
	    echo "#  include archlinux_macbookretina::$i" >> init.pp
	    echo >> init.pp
	else
	    echo "  $summary" >> init.pp
	    echo "  include archlinux_macbookretina::$i" >> init.pp
	    echo >> init.pp
	fi
done
echo "}" >> init.pp

popd > /dev/null


lib/facter/bios_and_system.rb
#
# bios_and_system.rb
#
# This fact provides more details about underlying platform in terms of
# BIOS and System Information data that can be taken from the DMI (Direct
# Media Interface) space.
#
# from <https://github.com/kwilczynski/facter-facts/blob/master/bios_and_system.rb>
#

require 'facter'
require 'facter/manufacturer'

query = {
  'BIOS [Ii]nformation' => [
    { 'Vendor:'           => 'bios_vendor'            },
    { 'Version:'          => 'bios_version'           },
    { 'Release Date:'     => 'bios_release_date'      },
    { 'ROM Size:'         => 'bios_rom_size'          },
    { 'BIOS Revision:'    => 'bios_revision'          },
    { 'Firmware Revision' => 'bios_firmware_revision' }
  ],
  '[Ss]ystem [Ii]nformation' => [
    { 'Manufacturer:'     => 'system_manufacturer'  },
    { 'Product Name:'     => 'system_product_name'  },
    { 'Serial Number:'    => 'system_serial_number' },
    { 'UUID:'             => 'system_uuid'          },
    { 'Family(?: Name)?:' => 'system_family_name'   }
  ]
}

# We call existing helper function to do the heavy-lifting ...
Facter::Manufacturer.dmi_find_system_info(query)

# vim: set ts=2 sw=2 et :
# encoding: utf-8


lib/facter/swapfile_offset.rb
#
# swapfile_offset.rb
#
# This fact provides the filesystem offset of /swapfile if it exists.
#

require 'facter'

if Facter.value(:kernel) == 'Linux' and FileTest.exists?("/swapfile")

  # output looks something like:
  #[root@jarvis boot]# filefrag -v /swapfile | head -10
  #Filesystem type is: ef53
  #File size of /swapfile is 4294967296 (1048576 blocks of 4096 bytes)
  # ext:     logical_offset:        physical_offset: length:   expected: flags:
  #   0:        0..       0:     509952..    509952:      1:
  #   1:        1..    2047:     509953..    511999:   2047:             unwritten
  #   2:     2048..    4095:     512000..    514047:   2048:             unwritten
  #   3:     4096..    6143:     514048..    516095:   2048:             unwritten
  #   4:     6144..    8191:     520192..    522239:   2048:     516096: unwritten
  #   5:     8192..   10239:     522240..    524287:   2048:             unwritten
  #   6:    10240..   12287:     778240..    780287:   2048:     524288: unwritten
  #

  # We work-around an issue in Facter #10278 by forcing locale settings ...
  ENV['LANG']   = 'POSIX'
  ENV['LC_ALL'] = 'POSIX'

  Facter::Util::Resolution.exec('/usr/bin/filefrag -v /swapfile').each_line do |line|
    # Remove bloat ...
    line.strip!

    # Skip header line ...
    next if not line.match(/^0:+/)

    # Parse line and retrieve offset...
    offset = line.split(' ')[3].strip.gsub(/\./, "")

    Facter.add('swapfile_offset') do
      confine :kernel => :linux
      setcode { offset }
    end

  end

end
# vim: set ts=2 sw=2 et :
# encoding: utf-8


lib/facter/local_volumes.rb
#
# local_volumes.rb
#
# This fact provides an alphabetic list of local volumes that are currently
# available and/or accessible on the systems.
#
# This fact was initially written by Matt Moran and this implementation is
# simply an improvement upon it.
#
# from <https://github.com/kwilczynski/facter-facts/blob/master/local_volumes.rb>
#

require 'thread'
require 'facter'

if Facter.value(:kernel) == 'Linux'
  mutex = Mutex.new

  # We store the list of local volumes here ...
  volumes = []
  partitions_sda = []

  # Modern Linux distributions provide "df -lP" output in the following format:
  #
  #   Filesystem         1024-blocks      Used Available Capacity Mounted on
  #   /dev/hda1              5232640   3995212   1237428      77% /
  #   udev                     10240       116     10124       2% /dev
  #   /dev/hda2             41922560  33766468   8156092      81% /home
  #   shm                    1048712         0   1048712       0% /dev/shm
  #
  # Given "-P" command line switch enables POSIX compatibility and ensures
  # consistent format for output.
  #

  # We work-around an issue in Facter #10278 by forcing locale settings ...
  ENV['LANG']   = 'POSIX'
  ENV['LC_ALL'] = 'POSIX'

  #
  # We utilise rely on "cat" for reading values from entries under "/proc".
  # This is due to some problems with IO#read in Ruby and reading content of
  # the "proc" file system that was reported more than once in the past ...
  #
  Facter::Util::Resolution.exec('/bin/df -lP').each_line do |line|
    # Remove bloat ...
    line.strip!

    # Skip header line ...
    next if line.match(/^[Ff]ile.+/)

    # Skip new and empty lines ...
    next if line.match(/^(\r\n|\n|\s*)$|^$/)

    # Parse line and retrieve name of the local volume ...
    volume = line.split(' ')[5].strip

    # Add single volume to the list ...
    mutex.synchronize { volumes << volume }

    fs = line.split(' ')[0].strip

    next if not fs.match(/^\/dev\/sda.+$/)

    mutex.synchronize { partitions_sda << fs }
  end

  Facter.add('local_volumes') do
    confine :kernel => :linux
    setcode { volumes.sort.join(',') }
  end

  Facter.add('partitions_sda') do
    confine :kernel => :linux
    setcode { partitions_sda.sort.join(',') }
  end
end

# vim: set ts=2 sw=2 et :
# encoding: utf-8
